---
title: "Analysis of RNAseq expressions of Eµ-Myc_dCas9a-SAM and B-All samples"
author: "Goknur Giner"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    code_folding: hide
  pdf_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, prompt=TRUE, warning=FALSE, comment=NA, message=FALSE, tidy = TRUE, size="footnotesize", fig.path = "./figures/", fig.height = 6)
```

# Backgroud
In this report, we explored the expression profiles of the genes that are differentially expressed between Eµ-Myc/dCas9a-SAM and B-All samples. The Eµ-Myc/dCas9a-SAM and B-ALL RNA-seq experiments are sequenced using Illumina NextSeq 2000 and HiSeq 2000 platforms, respectively, as paired-end 100 bp.
```{r loading pkgs and the data}
options(digits = 3)
rm(list = ls())
# setwd("~/Desktop/Eµ-Myc-dCas9a-SAM_RNASeq_Data")
setwd("/Volumes/GoogleDrive/My Drive/Professional/Manuscripts/before 2022/2021/Collaborations/SDiepstraten_CRISPRActivationScreen/Eµ-Myc-dCas9a-SAM_RNASeq_Data_RData")
library(edgeR)
library(limma)
library(gplots)
library(ggplot2)
library(pheatmap)
library(DT)
library(xlsx)
```
 
# Eµ-Myc/dCas9a-SAM data preparation
## Aligning and counting the samples
To align the Eµ-Myc/dCas9a-SAM samples and count the reads, we used Rsubread version 2.2.6 and BioConductor version 3.11.1. Below are the details.
```{r}
writeLines(readLines("./scripts/01_align.R"))
writeLines(readLines("./scripts/02_featurecount.R"))
```
## Checking the number of mapped reads 
Here we plot the number of successfully mapped reads and the proportion of mapped reads. As shown below, the proportion of successfully mapped reads are above 97% and the average library size is ~28 million. 
```{r mapping proportions, eval=FALSE}
load("./RData/Eµ-Myc-dCas9a-SAM_RNASeq.RData")
propmap <- out.gene$propmap
propmap$Name <- rownames(propmap)
ggplot(data = propmap, aes(x = reorder(Name, PropMapped), y = PropMapped)) +
  geom_bar(stat = "identity", fill = "#FF9999") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.95, vjust = 0.2)) +
  ggtitle("Proportion of mapped reads") +
  xlab("") +
  ylab("")
ggplot(data = propmap, aes(x = reorder(Name, NumMapped), y = NumMapped)) +
  geom_bar(stat = "identity", fill = "#FF9999") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.95, vjust = 0.2)) +
  ggtitle("Number of mapped reads") +
  xlab("") +
  ylab("")
```

## Annotation
Mouse genome annotation on the gene level is obtained from Rsubread and annotation for each Entrez Gene ID is obtained from the most up to date NCBI Mus_muculus.gene_info file.
```{r annotation, eval=FALSE}
d <- out.gene
ann <- d$count$annotation
ann$Chr <- unlist(lapply(strsplit(ann$Chr, ";"), function(x)
  paste(unique(x), collapse=",")))
ann$Strand <- ann$Start <- ann$End <- NULL
ann$Length <- ann$Length
ncbi.file <- "indices/Mus_musculus.gene_info"
ncbi.L1 <- readLines(ncbi.file, n = 1)
ncbi.colname <- unlist(strsplit(substring(ncbi.L1, 9, 240), '\t'))
ncbi <- read.delim(ncbi.file, skip = 1, head = F, stringsAsFactors = F, quote = "")
ncbi <- ncbi[,-1]
colnames(ncbi) <- ncbi.colname
m <- match(ann$GeneID, ncbi$GeneID)
ann$Symbol <- ncbi$Symbol[m]
ann$Type <- ncbi$type_of_gene[m]
ann$Description <- ncbi$description[m]
head(ann)
counts <- d$count$counts
d <- DGEList(counts, genes=ann[,c("Symbol","Chr", "Length", "Description", "Type")])
rownames(d$counts) <- d$genes$Symbol
dim(d)
```

## Creating the DGEList object
First we create a new DGEList object, which includes information about read counts for each gene, sample information and gene annotations. The new DGEList object has 27179 genes and 18 samples.
```{r dgelist, eval=FALSE}
d$samples$SampleNo <- strsplit2(rownames(d$samples),"_")[,1]
d <- d[,order(as.numeric(d$samples$SampleNo))]
Targets <- read.csv("Targets.csv")
d$samples$Type <- Targets$Type[1:18]
d$samples$Subtype <- Targets$Subtype[1:18]
d$samples$BioRep<- Targets$Biorep[1:18]
dim(d)
save(d,file="./RData/DHL_DGEList.RData")
```
# B-ALL data preparation
## Annotaion of B-ALL
```{r annotation b-all, eval=FALSE}
load("./RData/2022-05-03-B_ALL_RNA-Seq_from_Ashley-rsubread-mm10-gene.RData")
d <- out.gene
ann <- d$count$annotation
ann$Chr <- unlist(lapply(strsplit(ann$Chr, ";"), function(x)
  paste(unique(x), collapse=",")))
ann$Strand <- ann$Start <- ann$End <- NULL
ann$Length <- ann$Length
ncbi.file <- "indices/Mus_musculus.gene_info"
ncbi.L1 <- readLines(ncbi.file, n = 1)
ncbi.colname <- unlist(strsplit(substring(ncbi.L1, 9, 240), '\t'))
ncbi <- read.delim(ncbi.file, skip = 1, head = F, stringsAsFactors = F, quote = "")
ncbi <- ncbi[,-1]
colnames(ncbi) <- ncbi.colname
m <- match(ann$GeneID, ncbi$GeneID)
ann$Symbol <- ncbi$Symbol[m]
ann$Type <- ncbi$type_of_gene[m]
ann$Description <- ncbi$description[m]
head(ann)
counts <- d$count$counts
d <- DGEList(counts, genes=ann[,c("Symbol","Chr", "Length", "Description", "Type")])
rownames(d$counts) <- d$genes$Symbol
dim(d)
save(d,file="./RData/B-ALL_DGEList.RData")
```
# Combine Pre-B-ALL and Eμ-Myc/dCas9a-SAMKI/+/sgBcl-2 lymphoma counts in the same RData file
```{r combine dgeLists}
load("./RData/B-ALL_DGEList.RData")
m <- !is.na(d$genes$Symbol)
d.bll <- d[m,]
dim(d.bll)
d.bll
load("./RData/DHL_DGEList.RData")
m <- !is.na(d$genes$Symbol)
d.dhl <- d[m,]
d.dhl
sum(!rownames(d.dhl$genes)==rownames(d.bll$genes))
counts <- cbind(d.dhl$counts,d.bll$counts)
head(counts)
dim(counts)
d <- DGEList(counts, genes=d.dhl$genes)
rownames(d$genes) <- rownames(d.dhl$genes)
Targets <- read.csv("Targets.csv")
d$samples$Type <- Targets$Type
d$samples$Subtype <- Targets$Subtype
d$samples$BioRep<- Targets$Biorep
d$samples$Origin <- Targets$Origin
dim(d)
save(d,file="./RData/Combined_DGEList.RData")
dim(d)
```
# Figure 6a
```{r heatmaps}
load("./RData/Combined_DGEList.RData")
# Select Pre-B-ALL and Eμ-Myc/dCas9a-SAMKI/+/sgBcl-2 lymphoma samples
d <- d[,rownames(d$samples) %in% c("14_179_IgMp_S11_L001_R1_001.bam",
                                   "15_180_IgMp_S12_L001_R1_001.bam",
                                   "16_202_IgMp_S13_L001_R1_001.bam",
                                   "17_213_IgMp_S14_L001_R1_001.bam",
                                   "18_214_IgMn_S15_L001_R1_001.bam",
                                   "19_216_IgMn_S16_L001_R1_001.bam",
                                   "P190CreERLSLErg__138_C8HKNANXX_TAGCTT_L005_R1.fastq.gz.subjunc.BAM",
                                   "P190RagCreErg__99_C8HKNANXX_GGCTAC_L005_R1.fastq.gz.subjunc.BAM")]
dim(d)
d$samples$Subgroup <- "B-ALL"
d$samples$Subgroup[d$samples$Subtype=="IgM"] = "Eµ-Myc_dCas9a-SAM"

## Filter and normalise
i <- d$genes$Type %in% c("ncRNA","protein-coding","scRNA","snoRNA","snRNA")
d <- d[i,,keep.lib.size=FALSE]
i <- filterByExpr(d, group=d$samples$Subgroup)
d <- d[i,,keep.lib.size=FALSE]
dim(d)
d <- calcNormFactors(d,method="TMM")
d$samples

## DE analysis
group <- factor(d$samples$Subgroup) # B-ALL is the base group
design <- model.matrix(~group)
v <- voomLmFit(d, design,plot = TRUE)
fit <- eBayes(v,robust=TRUE)
summary(dt <- decideTests(fit))
toptable <- topTable(fit, n = Inf,coef = "groupEµ-Myc_dCas9a-SAM")
head(toptable, 20)
write.csv(toptable, paste0("tables/",Sys.Date(),"-DE_gene_list_groupEµ-Myc_dCas9a-SAM_vs_B-ALL_GGiner.csv"))

## Heatmap of top 200 DE genes
gene.list <- toptable[toptable$adj.P.Val < 0.05,]$Symbol
bp <- d$counts
rownames(bp) <- d$genes$Symbol
colnames(bp) <-d$samples$BioRep
dim(bp)
head(bp)
dim(bp <- bp[rowSums(bp)!=0,]) 
dim(bp <- bp[rownames(bp) %in% gene.list[1:200],])
m <- match(rownames(bp),gene.list[1:200])
# CPM normalisation + Row scaling for 3 main IG type

# excel table of genes with BCL2 and MYC translocations
genes_mono <- read.csv("./genes_BCL2_MYC.csv",header=FALSE)
write.csv(toptable[toptable$Symbol %in% genes_mono$V1,],"./tables/DE_results_for_genes_with_BCL2_or_MYC_translocations.csv")

# pathway analysis
Mm.c2 <- readRDS("./MSigDB/Mm.c2.all.v7.1.entrez.rds")
t.stats <- fit$t[,"groupEµ-Myc_dCas9a-SAM"]
names(t.stats) <- rownames(fit$genes)
camera.table <- cameraPR(t.stats,Mm.c2)
camera.table <- camera.table[camera.table$FDR<=0.05,]
write.xlsx(camera.table, 
           paste0("./tables/", Sys.Date(),"-MSigDB_Eµ-Myc_dCas9a-SAMvsB-ALL-GGiner.xlsx"))

pdf("./figures/Figure_6a.pdf", width = 10, height = 8)
H_res <- camera.table[grep("MYC|DLBCL", rownames(camera.table)),]
FDR = 0.05
h_plot <- H_res[H_res$FDR < FDR & H_res$Direction == "Up", ]
tmp <- H_res[H_res$FDR < FDR & H_res$Direction == "Down", ]
tmp <- tmp[nrow(tmp):1, ]
h_plot <- rbind(h_plot, tmp)
h_plot$FDR <- -log10(h_plot$FDR)
h_plot <- h_plot[nrow(h_plot):1, ]
h_cols <- as.factor(h_plot$Direction)
levels(h_cols) <- c(4, 2)
par(mar = c(5, 20, 4, 2))
barplot(h_plot$FDR, las = 2, horiz = TRUE, names.arg = rownames(h_plot), cex.names = 0.6, axes = TRUE, 
        border = 1, main = "Enriched gene sets in Eµ-Myc_dCas9a-SAM compared to B-ALL", xlab = "-log10(FDR)", col = as.character(h_cols))
abline(v = -log10(0.05), lty = 2, col = 1)
legend("right", c("Up", "Down"), pch = 15, col = c(2, 4))
dev.off()
```

# Supp Figure 3d
## Filtering and normalising
In this section, we remove the genes with no annotation and with low counts across all libraries.  After filtering, 14768 genes kept in the analysis.

```{r filtering}
load("./RData/Combined_DGEList.RData")
d <- d[,d$samples$Type == "MEFs"]
dim(d)
i <- d$genes$Type %in% c("ncRNA","protein-coding","scRNA","snoRNA","snRNA")
d <- d[i,,keep.lib.size=FALSE]
i <- filterByExpr(d, group=d$samples$Subtype)
d <- d[i,,keep.lib.size=FALSE]
dim(d)
```
After we filtered the low counts, we performed TMM normalisation to eliminate composition biases between libraries. 
```{r normalising}
d <- calcNormFactors(d,method="TMM")
d$samples
```

## Deciding the design of the experiment and Estimating dispersion and BCV plot
Next we define the design matrix based on the experimental design. Then we estimate the dispersion between biological replicates. 

```{r estimate dispersion}
Experiment <- factor(d$samples$Subtype, 
                     levels = c("WT",
                                "BCL2_1",
                                "BCL2_2",
                                "NTC",
                                "MEFs"))
design <- model.matrix(~Experiment)
d <- estimateDisp(d,design=design)
d$common.dispersion
```
The square root of the common dispersion gives the coefficient of variation of biological variation. Here the common dispersion is found to be 0.0688, so the coefficient of biological variation is around 0.262. The dispersion estimates can be viewed in a BCV plot:

```{r bcvplot}
plotBCV(d)
```

The biological coefficient of variation (BCV) plots show the BCV for each gene. 

## Differential expression analysis
Here we performed limma:voomLmFit pipeline in limma R package to conduct a differential expression analysis. 

## Groups compared to NTC
```{r voom}
Experiment <- factor(d$samples$Subtype, 
                     levels = c("NTC",
                                "BCL2_1",
                                "BCL2_2",
                                "WT",
                                "MEFs"))
design <- model.matrix(~Experiment)
v <- voomLmFit(d, design,plot = TRUE)
fit <- eBayes(v,robust=TRUE)
summary(dt <- decideTests(fit))
pdf("./figures/Sup_Figure_3d.pdf")
plotMD(fit, column=4, status=dt[,4], main="WT MEFs vs NTC")
abline(h=c(-1, 1), col="blue")
dev.off()
```

```{r session info}
sessionInfo()
```

# Figure 6d
```{r fig 6d}
load("./RData/Combined_DGEList.RData")
# Heatmap and pathway analysis with Ig negative samples
d <- d[,rownames(d$samples) %in% c("14_179_IgMp_S11_L001_R1_001.bam",
                                   "15_180_IgMp_S12_L001_R1_001.bam",
                                   "16_202_IgMp_S13_L001_R1_001.bam",
                                   "17_213_IgMp_S14_L001_R1_001.bam",
                                   "18_214_IgMn_S15_L001_R1_001.bam",
                                   "19_216_IgMn_S16_L001_R1_001.bam",
                                   "P190CreERLSLErg__138_C8HKNANXX_TAGCTT_L005_R1.fastq.gz.subjunc.BAM",
                                   "P190RagCreErg__99_C8HKNANXX_GGCTAC_L005_R1.fastq.gz.subjunc.BAM")]
dim(d)
d$samples$Subgroup[d$samples$Subtype=="B-ALL"] = "B-ALL"
d$samples$Subgroup[d$samples$Type=="IgMp"] = "DHL.IgM.positive"
d$samples$Subgroup[d$samples$Type=="IgMn"] = "DHL.IgM.negative"

## Filter and normalise
i <- d$genes$Type %in% c("ncRNA","protein-coding","scRNA","snoRNA","snRNA")
d <- d[i,,keep.lib.size=FALSE]
i <- filterByExpr(d, group=d$samples$Subgroup)
d <- d[i,,keep.lib.size=FALSE]
dim(d)
d <- calcNormFactors(d,method="TMM")
d$samples

## DE analysis
group <- factor(d$samples$Subgroup) # B-ALL is the base group
design <- model.matrix(~group)
v <- voomLmFit(d, design)
fit <- eBayes(v,robust=TRUE)
summary(dt <- decideTests(fit))
##########################
### IgMn vs B-ALL
##########################
toptable <- topTable(fit, n = Inf,coef = "groupDHL.IgM.negative")
head(toptable, 20)
write.csv(toptable, paste0(Sys.Date(),"-DE_gene_list_DHL_IgM_negative_vs_B-ALL_GGiner.csv"))

## Heatmap of DE genes
gene.list <- toptable[toptable$adj.P.Val < 0.05,]$Symbol
bp <- d$counts
# rownames(bp) <- substr(d.ig.exp$genes$Symbol,1,4)
rownames(bp) <- d$genes$Symbol
colnames(bp) <- paste0(d$samples$Subgroup,
                       "_", 
                       d$samples$BioRep)
dim(bp)
head(bp)
dim(bp <- bp[rowSums(bp)!=0,]) 
dim(bp <- bp[rownames(bp) %in% gene.list[1:200],])
m <- match(rownames(bp),gene.list[1:200])
# CPM normalisation + Row scaling for 3 main IG type
pdf("./figures/Figure_6d.pdf")
my_sample_col <- data.frame(Group=strsplit2(colnames(bp),split = "_")[,1],row.names = colnames(bp))
HMpal <- colorRampPalette(c(4, "white", 2))
pheatmap(bp, scale = "row", annotation_col = my_sample_col,show_rownames = TRUE, color=HMpal(256), main="Double Hit Lymphoma (Top 200 IgM- DE genes)",fontsize_row = 2)  
dev.off()
```

